src/ir/properties.cpp

```diff
@@ -36,4 +37,52 @@ bool isGenerative(Expression* curr, FeatureSet features) {
   return scanner.generative;
 }
 
+static bool isValidInConstantExpression(Module& wasm, Expression* expr) {
+  if (isSingleConstantExpression(expr) || expr->is<StructNew>() ||
+      expr->is<ArrayNew>() || expr->is<ArrayNewFixed>() || expr->is<I31New>() ||
+      expr->is<StringConst>()) {
+    return true;
+  }
+
+  if (auto* get = expr->dynCast<GlobalGet>()) {
+    auto* g = wasm.getGlobalOrNull(get->name);
+    if (!g) {
+      return false;
+    }
+    // Only gets of immutable globals are constant.
+    if (g->mutable_) {
+      return false;
+    }
+    // Only imported globals are available in constant expressions unless GC is
+    // enabled.
+    return g->imported() || wasm.features.hasGC();
+    // TODO: Check that there are no cycles between globals.
+  }
+
+  if (wasm.features.hasExtendedConst()) {
+    if (auto* bin = expr->dynCast<Binary>()) {
+      if (bin->op == AddInt64 || bin->op == SubInt64 || bin->op == MulInt64 ||
+          bin->op == AddInt32 || bin->op == SubInt32 || bin->op == MulInt32) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool isValidConstantExpression(Module& wasm, Expression* expr) {
+  std::vector<Expression*> exprs = {expr};
+  while (!exprs.empty()) {
+    Expression* curr = exprs.back();
+    exprs.pop_back();
+    if (!isValidInConstantExpression(wasm, curr)) {
+      return false;
+    }
+    ChildIterator children(curr);
+    exprs.insert(exprs.end(), children.begin(), children.end());
+  }
+  return true;
```

Comment:
We should really upgrade to C++20 and use generators so we can avoid all that boilerplate and imperative nonsense 😢

Reply:
There's a little boilerplate but the last 4 lines could be avoided if we added a default constructor in a helper class perhaps. Then it's really just 5 lines with code for the whole thing. I can look into that.

But I'm also curious to see what a generator solution would look like (and how fast it would be?)

Reply:
I looked into the constructor stuff, but the pattern is pretty rare and I can't find a good way to really shorten it. The problem is that adding a constructor in the parent class doesn't mean we can do `Child(module)` to create a child; we also need to manually define the constructor in the child. Unless there's some C++ magic?

Reply:
No, unfortunately there's no way around copying the constructor definition into every subclass.